### DISCUSSION TITLE
Migration ... failed to apply cleanly to the shadow database.  The  relation "..." already exists

### DISCUSSION THREAD

**Discussion by kravenhunter** [prisma/prisma#21401] [Q&A] [answered]
### Question

Bug description
Hello. I added a new model to prisma schema  and tried  to create  a new migration and got this error. 

```
npx prisma migrate dev --name changes                 
Environment variables loaded from .env
Prisma schema loaded from prisma\schema.prisma
Datasource "db": PostgreSQL database "postgres", schema "public" at "aws-0-eu-central-1.pooler.supabase.com:5432"    

Error: P3006

Migration `20231004180313_init` failed to apply cleanly to the shadow database.
Error:
db error: ERROR: relation "PageContents" already exists
   0: schema_core::state::DevDiagnostic
             at schema-engine\core\src\state.rs:270
```
But first migration had created fine.

The error occure when I add the Contacts model

### New model
```prisma
model Contacts {
  id                String      @id @default(uuid())
  timeStamp         DateTime    @default(now())
  imageBgLink        String
  imagePreviewLink   String? 
  title             String     
  shortDescription String?
  description_1    String?
  description_2    String?
  description_3    String?
  description_4    String?
  phone             String?
  mobile           String?
  email             String?
  socialLink_1       String?
  socialLink_2       String?
  socialLink_3       String?
  socialLink_4       String?
  socialLink_5       String?
  copyright          String? 
     
}
```

### Expected behavior (optional)

I expected opportunity to create new tables  in my database when some tables in db allready have filled  without dropping all tables,migration  adn create it again

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}



model PageContents {
  id               String   @id @default(uuid())
  timeStamp        DateTime @default(now())
  imageBgLink      String
  imagePreviewLink String?
  galary_link_1    String?
  galary_link_2    String?
  galary_link_3    String?
  galary_link_4    String?
  title            String
  subTitle         String
  shortDescription String
  description_1    String
  description_2    String?
  description_3    String?
  description_4    String?
}

model UserProfile {  
  id         String   @id  @unique
  email       String
  timeStamp   DateTime @default(now())
  date_of_birth  String
  phone      String
  login_name  String
  first_name String
  last_name  String
  country    String
  adress     String
  avatar     String
  avatarPreview String
  type       UserTypeStatus

}

model Basket {
  id         String @id @default(uuid())
 timeStamp        DateTime @default(now())
  product_id String
  count      Int
  user_id    String
}

model Product {
  id               String   @id @default(uuid())
  timeStamp        DateTime @default(now())
  imageBgLink      String
  imagePreviewLink String?
  galary_link_1    String
  galary_link_2    String
  galary_link_3    String
  galary_link_4    String
  likes            Int
  dislikes         Int
  title            String
  subTitle         String
  category         Category
  price            Decimal
  productCode      Decimal
  discont          Int
  size             String
  weight           Decimal
  dimension        String
  dough            String
  calories         Decimal
  fat              Decimal
  arbohydrates     Decimal
  carbohydrates    Decimal
  shortDescription String
  description_1    String
  description_2    String?
  description_3    String?
  description_4    String?
}
enum UserTypeStatus {
  Administrator
  Moderator
  User
}

enum Category {
  Pizza
  Burger
  Salad
  Soup
  Sushi
  Dessert
  Drinks
  First
  Second
  FastFood
  Additionals
}`
```

### Prisma Client queries

```ts

CREATE TABLE "PageContents" (
    "id" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "imageBgLink" TEXT NOT NULL,
    "imagePreviewLink" TEXT,
    "galary_link_1" TEXT,
    "galary_link_2" TEXT,
    "galary_link_3" TEXT,
    "galary_link_4" TEXT,
    "title" TEXT NOT NULL,
    "subTitle" TEXT NOT NULL,
    "shortDescription" TEXT NOT NULL,
    "description_1" TEXT NOT NULL,
    "description_2" TEXT,
    "description_3" TEXT,
    "description_4" TEXT,

    CONSTRAINT "PageContents_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "UserProfile" (
    "id" TEXT NOT NULL,
    "email" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "date_of_birth" TEXT NOT NULL,
    "phone" TEXT NOT NULL,
    "login_name" TEXT NOT NULL,
    "first_name" TEXT NOT NULL,
    "last_name" TEXT NOT NULL,
    "country" TEXT NOT NULL,
    "adress" TEXT NOT NULL,
    "avatar" TEXT NOT NULL,
    "avatarPreview" TEXT NOT NULL,
    "type" "UserTypeStatus" NOT NULL,

    CONSTRAINT "UserProfile_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "Basket" (
    "id" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "product_id" TEXT NOT NULL,
    "count" INTEGER NOT NULL,
    "user_id" TEXT NOT NULL,

    CONSTRAINT "Basket_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "Product" (
    "id" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "imageBgLink" TEXT NOT NULL,
    "imagePreviewLink" TEXT,
    "galary_link_1" TEXT NOT NULL,
    "galary_link_2" TEXT NOT NULL,
    "galary_link_3" TEXT NOT NULL,
    "galary_link_4" TEXT NOT NULL,
    "likes" INTEGER NOT NULL,
    "dislikes" INTEGER NOT NULL,
    "title" TEXT NOT NULL,
    "subTitle" TEXT NOT NULL,
    "category" "Category" NOT NULL,
    "price" DECIMAL(65,30) NOT NULL,
    "productCode" DECIMAL(65,30) NOT NULL,
    "discont" INTEGER NOT NULL,
    "size" TEXT NOT NULL,
    "weight" DECIMAL(65,30) NOT NULL,
    "dimension" TEXT NOT NULL,
    "dough" TEXT NOT NULL,
    "calories" DECIMAL(65,30) NOT NULL,
    "fat" DECIMAL(65,30) NOT NULL,
    "arbohydrates" DECIMAL(65,30) NOT NULL,
    "carbohydrates" DECIMAL(65,30) NOT NULL,
    "shortDescription" TEXT NOT NULL,
    "description_1" TEXT NOT NULL,
    "description_2" TEXT,
    "description_3" TEXT,
    "description_4" TEXT,

    CONSTRAINT "Product_pkey" PRIMARY KEY ("id")
);

-- CreateIndex
CREATE UNIQUE INDEX "UserProfile_id_key" ON "UserProfile"("id");

```

- OS:  Windows 11
- I use Supabase service  with PostgreSQL Database
- Node.js version: v18.16.0

- Prisma and Prisma Client version 5.3.1

**Comment by ludralph** [comment]
Hi @kravenhunter 👋 

Thank you for raising this question.

You are getting this error because the state of your local development database and Prisma schema are not in sync.

You can try resetting the database using the `prisma migrate reset` command and then try reapplying the migrations again using the `prisma migrate dev` command again.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by kravenhunter** [reply]
Thanks for you answer.  I've just read  that the prisma migrate reset command can drop my database and it's not  that way  I'm looking for, because  I already have  filled database in postgresql. Is there any way to add new table without deleting all my data or  I'll have to transfer my data to seed file  and recreate database  by using the reset command ?

**Reply by janpio** [reply]
Something is off with your migration history if it tries to create a table that already exists.
You will need to figure out how to fix that. Can you shared your migration files that match this schema?

**Reply by kravenhunter** [reply]
I couldn't find hot to attach the prisma folder to my message so here's my link to repo https://github.com/kravenhunter/pizza

**Reply by janpio** [reply]
That migration looks clean. It creates all the tables from nothing. As this is first executed on a shadow database (a new temporary database, which is also mentioned in the error message), this should succeed. Did you try again?

**Reply by kravenhunter** [reply]
Yes,  by now it first migration. But before  , when  the error  had happened I had to delete  old migration and all data in database and recreate migration again   that to add new table to database . And now  I  needed to add the Contact model and  error appeared again.  I tried  many time  run   the command  " npx prisma migrate dev --name changes" , but I'm  still  receiving  the error :

`Error: P3006

Migration `20231004180313_init` failed to apply cleanly to the shadow database.
Error:
db error: ERROR: relation "PageContents" already exists
   0: schema_core::state::DevDiagnostic
             at schema-engine\core\src\state.rs:270`

And I don't undestrand does prisma allow  to create tables when database  isn't empty or I have to erase all my data every time and   recreate migration that to add new table?

**Reply by janpio** [reply]
No, Prisma should create a new empty shadow database in the background that is empty, then apply all your migration history to it, so it can calculate the requires SQL diff that it needs to put in your new migration file to get to the state of the database you are aiming for. You are doing something in an unexpected way then I would say.

Can you recap _exactly_ the steps and commands you are running?

One thing: You should not be using the pooled connection for `migrate dev` and similar commands. Does it go away when you use the direct connection string? (THe `pooler` in the connection indicates you might not be doing that right now).

**Reply by kravenhunter** [reply]
I've just followed this steps:
1. Installed  prisma and run " npx prisma  init " . 
2. Created first migration "npx prisma migrate dev --name init" 
3. And whet I needed to add a new table I added a new model to the prisma schema and  tried  to run "npx prisma migrate dev --name changes" 
4. To format prisma schema  used "npx prisma format" 
That's all command I used

**Reply by janpio** [reply]
1. Create database on Supabase
2. Wait for it to be created
3. Get connection string from "Project Settings" => "Database" => "Connection String" => "NodeJS": `postgresql://postgres:[YOUR-PASSWORD]@db.xpkvpofuupyyhktgkmpg.supabase.co:5432/postgres` 
4. `mkdir 21401 && cd 21401 && npm init -y && npm install prisma && npx prisma init`
5. Set connection string in `.env`
6. Add minimal model in `schema.prisma`: 
	```
	model User {
	  id   String  @id @default(uuid())
	  name String?
	}
	```
7. `migrate dev --name init` to generate SQL and deploy:
	```sql
	-- CreateTable
	CREATE TABLE "User" (
	    "id" TEXT NOT NULL,
	    "name" TEXT,
	
	    CONSTRAINT "User_pkey" PRIMARY KEY ("id")
	);
	```
8. Edit schema minimally to add another column:
9. `migrate dev --name changes` to generate and deploy, get a warning:
	```
	⚠️  Warnings for the current datasource:
	
	  • A unique constraint covering the columns `[email]` on the table `User` will be added. If there are existing duplicate values, this will fail.
	```
	But then:
	```sql
	/*
	  Warnings:
	
	  - A unique constraint covering the columns `[email]` on the table `User` will be added. If there are existing duplicate values, this will fail.
	  - Added the required column `email` to the `User` table without a default value. This is not possible if the table is not empty.
	
	*/
	-- AlterTable
	ALTER TABLE "User" ADD COLUMN     "email" TEXT NOT NULL;
	
	-- CreateIndex
	CREATE UNIQUE INDEX "User_email_key" ON "User"("email");
	```
10. And I have everything properly applied and migrated.

**Reply by janpio** [reply]
But that also highlights again what I wrote before and you did not acknowledge:

> You should not be using the pooled connection for migrate dev and similar commands. Does it go away when you use the direct connection string? (The `pooler` in the connection indicates you might not be doing that right now).

Using the pooled connection I get unexpected errors.

**Reply by kravenhunter** [reply]
I'm sorry that I didn't answer about the pooler connection  because  at first   didn't get  what you meant. But  yes, you was right. I used  the  Pooling Custom Configuration instead of   the NodeJS connection.  So ,I 've  change  it to the  NodeJS connection string and now I'm  getting a new error:
```
Error: P3006
Migration `20231004180313_init` failed to apply cleanly to the shadow database.
Error:
db error: ERROR: type "UserTypeStatus" does not exist
   0: schema_core::state::DevDiagnostic
             at schema-engine\core\src\state.rs:270
```

That's strange , because  I have  the UserProfile model created in database and  which uses the  UserTypeStatus  enum type. 
``` 
-- CreateTable
CREATE TABLE "UserProfile" (
    "id" TEXT NOT NULL,
    "email" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "date_of_birth" TEXT NOT NULL,
    "phone" TEXT NOT NULL,
    "login_name" TEXT NOT NULL,
    "first_name" TEXT NOT NULL,
    "last_name" TEXT NOT NULL,
    "country" TEXT NOT NULL,
    "adress" TEXT NOT NULL,
    "avatar" TEXT NOT NULL,
    "avatarPreview" TEXT NOT NULL,
    "type" "UserTypeStatus" NOT NULL,

    CONSTRAINT "UserProfile_pkey" PRIMARY KEY ("id")
);
``` 
I've created a user in  the UserProfile table and tried  run migration again  but  it didn.'t worked, the same error , that  type "UserTypeStatus" does not exist.

**Reply by janpio** [reply]
In the migration file that is committed to your repository it indeed uses the `UserTypeStatus` enum type, but does not actually create it:
https://github.com/kravenhunter/pizza/blob/fae0bc9959709c97a0c98dc7d264426094f64f03/prisma/migrations/20231004180313_init/migration.sql#L36C13-L36C27

It should actually look like this:
```sql
-- CreateEnum
CREATE TYPE "UserTypeStatus" AS ENUM ('Administrator', 'Moderator', 'User');

-- CreateEnum
CREATE TYPE "Category" AS ENUM ('Pizza', 'Burger', 'Salad', 'Soup', 'Sushi', 'Dessert', 'Drinks', 'First', 'Second', 'FastFood', 'Additionals');

-- CreateTable
CREATE TABLE "PageContents" (
    "id" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "imageBgLink" TEXT NOT NULL,
    "imagePreviewLink" TEXT,
    "galary_link_1" TEXT,
    "galary_link_2" TEXT,
    "galary_link_3" TEXT,
    "galary_link_4" TEXT,
    "title" TEXT NOT NULL,
    "subTitle" TEXT NOT NULL,
    "shortDescription" TEXT NOT NULL,
    "description_1" TEXT NOT NULL,
    "description_2" TEXT,
    "description_3" TEXT,
    "description_4" TEXT,

    CONSTRAINT "PageContents_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "UserProfile" (
    "id" TEXT NOT NULL,
    "email" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "date_of_birth" TEXT NOT NULL,
    "phone" TEXT NOT NULL,
    "login_name" TEXT NOT NULL,
    "first_name" TEXT NOT NULL,
    "last_name" TEXT NOT NULL,
    "country" TEXT NOT NULL,
    "adress" TEXT NOT NULL,
    "avatar" TEXT NOT NULL,
    "avatarPreview" TEXT NOT NULL,
    "type" "UserTypeStatus" NOT NULL,

    CONSTRAINT "UserProfile_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "Basket" (
    "id" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "product_id" TEXT NOT NULL,
    "count" INTEGER NOT NULL,
    "user_id" TEXT NOT NULL,

    CONSTRAINT "Basket_pkey" PRIMARY KEY ("id")
);

-- CreateTable
CREATE TABLE "Product" (
    "id" TEXT NOT NULL,
    "timeStamp" TIMESTAMP(3) NOT NULL DEFAULT CURRENT_TIMESTAMP,
    "imageBgLink" TEXT NOT NULL,
    "imagePreviewLink" TEXT,
    "galary_link_1" TEXT NOT NULL,
    "galary_link_2" TEXT NOT NULL,
    "galary_link_3" TEXT NOT NULL,
    "galary_link_4" TEXT NOT NULL,
    "likes" INTEGER NOT NULL,
    "dislikes" INTEGER NOT NULL,
    "title" TEXT NOT NULL,
    "subTitle" TEXT NOT NULL,
    "category" "Category" NOT NULL,
    "price" DECIMAL(65,30) NOT NULL,
    "productCode" DECIMAL(65,30) NOT NULL,
    "discont" INTEGER NOT NULL,
    "size" TEXT NOT NULL,
    "weight" DECIMAL(65,30) NOT NULL,
    "dimension" TEXT NOT NULL,
    "dough" TEXT NOT NULL,
    "calories" DECIMAL(65,30) NOT NULL,
    "fat" DECIMAL(65,30) NOT NULL,
    "arbohydrates" DECIMAL(65,30) NOT NULL,
    "carbohydrates" DECIMAL(65,30) NOT NULL,
    "shortDescription" TEXT NOT NULL,
    "description_1" TEXT NOT NULL,
    "description_2" TEXT,
    "description_3" TEXT,
    "description_4" TEXT,

    CONSTRAINT "Product_pkey" PRIMARY KEY ("id")
);

-- CreateIndex
CREATE UNIQUE INDEX "UserProfile_id_key" ON "UserProfile"("id");
```

Did you modify the migration file manually?

**Reply by kravenhunter** [reply]
Absolutely not. I only worked with prisma schema. And It looks strange  because  I can add  Product records and UserProfile records to the  database  that contains Category enum and UserTypeStatus enum. and I don't recieve any error. I guess  the Category and UserTypeStatus   somehow had been created  in the database but didn't added to the migration file. And I checked   a timestamp  of the migration file   and  the created time matches to  the change time  till seconds, so the migration file  definitely  didn't edit.

**Reply by janpio** [reply, marked as answer]
Can you start over, create a new database and delete the migration file to see if you can reproduce this again?

**Reply by kravenhunter** [reply]
It's finally started working. I saved my data , then I add  2 strings to migration file where creates  UserTypeStatus and Category enums. Then  I tried to run  command "npx prisma migrate dev --name changes" and the prisma ran reset command and  delete all my tables ,then 
 it recreated the migration and tables. So, now I can add new models and create new migration without any error.   @janpio @ludralph   thank you very much for your help and your detailed answers!!

**Reply by janpio** [reply]
Awesome. 

Let us know when you figure out what went wrong, maybe there is a way we avoid that in the future for other people.

**Reply by johannbuscail** [reply]
@kravenhunter can you explain how you solved it ? I'm facing the same problem here.

**Reply by kravenhunter** [reply]
@johannbuscail  If youre using Supabase then go to your supabase accout and copy the NodeJS connection string instead . of using Pooling Custom Configuration . Then run next command npm prisma migrate reset and try to create a new migration npm prisma migrate dev -name

**Comment by shoyeb001** [comment]
This error occurs when you delete you previous migration folders or some of the migration folders are missing. So you can manage the previous migration files and run `npx prisma migrate reset.`  But if you dont get the previous folder then delete all the migrations and run `npx prisma migrate reset `and then `npx prisma migrate dev --name init`.

**Reply by temtechie** [reply]
@shoyeb001 

This command `npx prisma migrate reset` is saying all data will be lost, I can not afford to loose my data


error message:
` npx prisma migrate reset
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "id8r" at "id8r.cx2kiwgkrbam.us-east-1.rds.amazonaws.com:3306"

✔ Are you sure you want to reset your database? All data will be lost. … no`

How do i go about it?

**Reply by SySagar** [reply]
did you find any alternative?
